File: app/src/main/kotlin/tech/mmarca/openvitals/data/repository/contract/CycleRepository.kt
package tech.mmarca.openvitals.data.repository.contract

import java.time.LocalDate
import tech.mmarca.openvitals.core.period.PeriodLoadQuery
import tech.mmarca.openvitals.domain.cycle.CycleStatistics
import tech.mmarca.openvitals.domain.model.CycleData
import tech.mmarca.openvitals.domain.model.CycleEntry
import tech.mmarca.openvitals.domain.model.CycleEntryKind
import tech.mmarca.openvitals.domain.model.CycleEntryWriteRequest
import tech.mmarca.openvitals.domain.model.RefreshMode
import tech.mmarca.openvitals.domain.query.CyclePeriodData

interface CycleRepository {
    val phase4Permissions: Set<String>

    suspend fun missingPermissions(): Set<String>

    suspend fun loadCyclePeriod(
        query: PeriodLoadQuery,
        refreshMode: RefreshMode = RefreshMode.NORMAL,
    ): CyclePeriodData

    suspend fun loadCycleData(start: LocalDate, end: LocalDate): CycleData

    fun cycleWritePermissions(kind: CycleEntryKind): Set<String>

    suspend fun hasCycleWritePermission(kind: CycleEntryKind): Boolean

    suspend fun writeCycleEntry(request: CycleEntryWriteRequest): String

    suspend fun loadCycleEntry(kind: CycleEntryKind, id: String): CycleEntry?

    suspend fun updateCycleEntry(id: String, request: CycleEntryWriteRequest)

    suspend fun deleteCycleEntry(kind: CycleEntryKind, id: String)

    suspend fun loadCycleStatistics(today: LocalDate = LocalDate.now()): CycleStatistics?
}

File: lib/data/repository/contract/cycle_repository.dart
import '../../../core/period/period_load_query.dart';
import '../../../core/result/result.dart';
import '../../../core/time/local_date.dart';
import '../../../domain/model/cycle_models.dart';
import '../../../domain/model/refresh_mode.dart';
import '../../../domain/query/cycle_period_data.dart';

/// Port of the Kotlin `CycleRepository` contract.
///
/// Fallible operations return [Result]; the synchronous probe
/// ([phase4Permissions]) reads cached state and cannot fail, so it stays bare.
abstract interface class CycleRepository {
  Set<String> get phase4Permissions;

  Future<Result<Set<String>>> missingPermissions();

  Future<Result<CyclePeriodData>> loadCyclePeriod(
    PeriodLoadQuery query, {
    RefreshMode refreshMode = RefreshMode.normal,
  });

  Future<Result<CycleData>> loadCycleData(LocalDate start, LocalDate end);
}
